src/types/widget.ts

```diff
@@ -31,7 +31,45 @@ interface DisposeOptions {
 }
 
 export type UiState = {
-  [stateKey: string]: any;
+  query?: string;
+  menu?: {
```

Comment:
I'm not sure this should be hardcoded though 🤔

Reply:
Would be more doing something should be imported from the widgets instead

Reply:
Do you mean that the type should be extended when a widget is added to InstantSearch?

Reply:
what I mainly mean is that this type could easily be "lying" if there's new widgets or change in the things it exposes, this will easily be missed in this type. Maybe it needs tests

Reply:
The type won't be "lying" because it's checked in each `getWidgetState()` connector method:

https://github.com/algolia/instantsearch.js/blob/adf8220064dd46b5b6d6b06ef0e858bf476f219b/src/connectors/infinite-hits/connectInfiniteHits.ts#L226-L237

If you were to create another widget, leading to the creation of a new key in the UiState, this would be easily spotted.

Reply:
@Haroenv could you help me understand what you mean? This is what I think. Let me know if I'm wrong.

If there's new widgets or change in already existing widgets, when we try to read/write something to/from uiState, the editor will show type error message. Then we realise the type is outdated and can update it to whatever we want it to be.

Reply:
good for me then I guess, but what about custom widgets in TS (later) who can't extend this type?

Reply:
That's a good point. When user creates a custom widget in their own project, can they extend this type, add their things, and overwrite it somehow (like in vue-shims.d.ts) ?

Reply:
That would be a good way to handle this in the future, indeed!

For now this PR is still an improvement over what we have right now and makes our data flow easier to understand.

Reply:
Right. We can proceed like this, and in the future users can extend this type if they build their own custom widgets in their code base like [this](https://stackoverflow.com/questions/43952198/how-to-override-or-extend-a-libary-type-definition-in-typescript).